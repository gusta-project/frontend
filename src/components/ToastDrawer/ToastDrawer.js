import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import PropTypes from 'prop-types';
import React from 'react';
import { Toast } from 'react-bootstrap';
import { useSelector } from 'react-redux';

import { getQueue } from 'selectors/toast';

export default function ToastDrawer() {
  const toasts = useSelector(getQueue);

  if (!Array.isArray(toasts) || toasts.length === 0) {
    return null;
  }

  return (
    <section className="toast-drawer">
      {toasts.map((toast) => {
        const { id, title, message, icon, show } = toast;

        return (
          <Toast key={id} show={show}>
            <Toast.Header>
              {icon && <FontAwesomeIcon icon={icon} />}
              <strong className="mr-auto">{title}</strong>
            </Toast.Header>
            <Toast.Body>{message}</Toast.Body>
          </Toast>
        );
      })}
    </section>
  );
}

ToastDrawer.propTypes = {
  toasts: PropTypes.arrayOf(
    PropTypes.shape({
      icon: PropTypes.string,
      interval: PropTypes.number,
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      message: PropTypes.string.isRequired
    })
  )
};
