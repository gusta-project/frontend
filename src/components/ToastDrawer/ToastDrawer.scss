.toast-drawer {
  position: absolute;
  right: 0;
  z-index: 100;
  padding-right: 40px;
}

.toast-header {
  @include themed {
    background-color: theme("toast-bg");
    color: theme("toast-fg");
  }
}

.toast {
  @include themed {
    background-color: theme("toast-bg");
    color: theme("toast-fg");
  }
}

.close {
  @include themed {
    color: theme("toast-close");
    text-shadow: theme("white-themed") 0 1px 0;
  }
}
