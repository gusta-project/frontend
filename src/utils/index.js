import dayjs from 'dayjs';

import { getApplication, getAuthorization } from 'selectors/application';

/* eslint-disable react/display-name, react/prop-types */

/**
 * Creates an object containing action constants namespaced under the specified reducer.
 *
 * @param {string} reducer The name of the reducer or module
 * @param {string[]} actions A list of action names
 * @return {object} Object with action name keys and full action string values
 */
export function buildActions(reducer, actions) {
  const result = {};

  for (const action of actions) {
    result[action] = `${reducer}/${action}`;
  }

  return result;
}

/**
 * Interpolate strings from the `params` object into the `url`.
 *
 * @param {object} endpoint Object with `url` and optional `params` object
 */
export const buildUrl = (endpoint) => {
  // this is defined by webpack.DefinePlugin
  const baseUrl = API_URL;
  const { url, params } = endpoint;

  if (!params) {
    return `${baseUrl}${url}`;
  }

  return Object.entries(params).reduce((resultUrl, entry) => {
    const [key, value] = entry;

    return resultUrl.replace(`{${key}}`, value);
  }, `${baseUrl}${url}`);
};

/**
 * Gets the initial Redux state from localStorage and the source code defaults.
 */
export const getInitialState = () => {
  // this needs to be require()d because an import results in a
  // circular dependency
  const { initialState } = require('reducers');

  try {
    const [accessToken, expiration] = [
      localStorage.getItem('accessToken'),
      localStorage.getItem('expiration')
    ];

    if (!accessToken) {
      return initialState;
    }

    const expirationDate = dayjs(JSON.parse(expiration));

    if (!expirationDate.isValid() || expirationDate.isBefore(dayjs())) {
      localStorage.removeItem('accessToken');
      localStorage.removeItem('expiration');
      return initialState;
    }

    return {
      ...initialState,
      application: {
        ...getApplication(initialState),
        authorization: {
          ...getAuthorization(initialState),
          accessToken: JSON.parse(accessToken),
          expiration: expirationDate
        }
      }
    };
  } catch {
    return initialState;
  }
};
