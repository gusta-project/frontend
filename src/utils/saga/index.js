import counterUtilities from './counter';
import pageUtilities from './pager';
import toastUtilities from './toasts';

export default {
  ...counterUtilities,
  ...pageUtilities,
  ...toastUtilities
};
