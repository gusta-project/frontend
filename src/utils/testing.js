import React, { createElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

/**
 * Creates a mock component which takes a set of props specified at runtime.
 *
 * @param {string} name The component/element name (in PascalCase)
 * @param {object} props An object containing props to setup
 * @return {object} Mock component with specified name and props
 */
export const mockComponent = (name, props = {}) => () =>
  createElement(name, props, props.children);

/**
 * Creates a mock component which will expose its props for snapshot testing purposes
 * @param {*} name The component/element name (in PascalCase)
 */
export const mockComponentWithProps = (name) => (props) =>
  createElement(name, props, props.children);

/**
 * Wraps a React component in a <MemoryRouter> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 */
export const withMemoryRouter = (WrappedComponent, routerProps = {}) => (
  props
) => (
  <MemoryRouter {...routerProps}>
    <WrappedComponent {...props} />
  </MemoryRouter>
);

/**
 * Wraps a React component in a <Provider> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 * @param {object} store Mock store
 */
export const withProvider = (WrappedComponent, store) => (props) => (
  <Provider store={store}>
    <WrappedComponent {...props} />
  </Provider>
);
