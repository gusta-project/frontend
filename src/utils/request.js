import axios from 'axios';
import { buildUrl } from 'utils';
import { call, race, delay } from 'redux-saga/effects';

/**
 * Make an XMLHttpRequest to the specified endpoint with the specified request data.
 *
 * @param {object} endpoint An object with `url` and `method` keys
 * @param {object} data An object to use as the request body
 */
class Request {
  *execute({ endpoint, data, options = {} }) {
    try {
      // ensure the endpoint was supplied
      if (!endpoint) {
        throw new Error('No endpoint provided!');
      }

      // ensure properties required by axios are present
      const { url, method } = endpoint;

      if (!url) {
        throw new Error('Endpoint is missing URL!');
      }

      if (!method) {
        throw new Error('Endpoint is missing method!');
      }

      // default timeout of 30 seconds
      const { timeout = 30000 } = options;
      const requestUrl = yield call(buildUrl, endpoint);

      // start a race between the request and a timer, cancel the loser
      const { response } = yield race({
        response: call(axios, {
          url: requestUrl,
          method,
          data
        }),
        timeout: delay(timeout)
      });

      if (!response) {
        throw new Error('Request timed out!');
      }

      const { status } = response;
      const success = status >= 200 && status < 400;

      if (success) {
        return {
          success,
          response
        };
      } else {
        return {
          success,
          error: response
        };
      }
    } catch (error) {
      return {
        success: false,
        error
      };
    }
  }
}

const instance = new Request();

export default instance;
