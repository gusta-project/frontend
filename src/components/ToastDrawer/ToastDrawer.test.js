import { render } from '@testing-library/react';
import React from 'react';
import configureStore from 'redux-mock-store';

import ToastDrawer from './ToastDrawer';
import { withProvider } from 'utils/testing';

describe('<ToastDrawer />', () => {
  const mockStore = configureStore();

  it('renders correctly', () => {
    const initialState = {
      toast: {
        queue: [
          {
            id: 'test',
            title: 'I am a toast',
            message: 'Ask me anything'
          }
        ]
      }
    };
    const store = mockStore(initialState);
    const ConnectedToastDrawer = withProvider(ToastDrawer, store);
    const { asFragment } = render(<ConnectedToastDrawer />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders null if empty', () => {
    const emptyState = { toast: { queue: [] } };
    const store = mockStore(emptyState);
    const ConnectedToastDrawer = withProvider(ToastDrawer, store);
    const { asFragment } = render(<ConnectedToastDrawer />);

    expect(asFragment()).toMatchSnapshot();
  });
});
