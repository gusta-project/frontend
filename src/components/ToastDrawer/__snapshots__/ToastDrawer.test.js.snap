// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToastDrawer /> renders correctly 1`] = `
<DocumentFragment>
  <section
    class="toast-drawer"
  >
    <div
      aria-atomic="true"
      aria-live="assertive"
      class="fade toast show"
      role="alert"
    >
      <div
        class="toast-header"
      >
        <strong
          class="mr-auto"
        >
          I am a toast
        </strong>
        <button
          class="close ml-2 mb-1"
          data-dismiss="toast"
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Ã—
          </span>
          <span
            class="sr-only"
          >
            Close
          </span>
        </button>
      </div>
      <div
        class="toast-body"
      >
        Ask me anything
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`<ToastDrawer /> renders null if empty 1`] = `<DocumentFragment />`;
